opte/src/engine/packet.rs

```diff
@@ -751,19 +736,124 @@ impl Packet<Initialized> {
+        // inclusive upper bound of the squash.
+        let squash_to = match (body.seg_index, body.seg_offset) {
+            // The body is in the first segment meaning all headers are also in
+            // the first segment. No squashing needed.
+            (0, _) => 0,
+
+            // The body starts at a zero offset in segment n. This means we need
+            // to squash all segments prior to n.
+            (n, 0) => n - 1,
+
+            // The body starts at a non-zero offset in segment n. This means we
+            // need to squash all segments up to and including n.
+            (n, _) => n,
+        };
+
+        // If the squash bound is zero, there is nothing left to do here, just
+        // return.
+        if squash_to == 0 {
+            return Ok(Packet {
+                avail: self.avail,
+                // The new packet is taking ownership of the segments.
+                segs: core::mem::take(&mut self.segs),
+                state: Parsed {
+                    len: self.state.len,
+                    hdr_offsets: info.offsets,
+                    meta: info.meta,
+                    flow,
+                    body_csum: info.body_csum,
+                    body,
+                    body_modified: false,
+                },
+            });
+        }
+
+        // Calculate the body offset within the new squashed segment
+        if body.seg_offset != 0 {
+            for s in &self.segs[..squash_to] {
+                body.seg_offset += s.len;
+            }
+        }
+        body.seg_index -= squash_to;
+
+        // Determine how big the message block for the squashed segment needs to
+        // be.
+        let mut new_seg_size = 0;
+        for s in &self.segs[..squash_to + 1] {
+            new_seg_size += s.len;
+        }
+
+        let extra_space = info.extra_hdr_space.unwrap_or(0);
+        let mut mp = allocb(new_seg_size + extra_space);
+        unsafe {
+            (*mp).b_wptr = (*mp).b_wptr.add(extra_space);
+            (*mp).b_rptr = (*mp).b_rptr.add(extra_space);
+            for s in &self.segs[..squash_to + 1] {
+                core::ptr::copy_nonoverlapping(
+                    (*s.mp).b_rptr,
+                    (*mp).b_wptr,
+                    s.len,
+                );
+                (*mp).b_wptr = (*mp).b_wptr.add(s.len);
+            }
+        }
+
+        // Construct a new segment vector, tacking on any remaining segments
+        // after the header segments.
+        let orig_segs = core::mem::take(&mut self.segs);
+        let mut segs = vec![unsafe { PacketSeg::wrap_mblk(mp) }];
+        if squash_to + 1 < orig_segs.len() {
+            segs[0].link(&orig_segs[squash_to + 1]);
+            segs.extend_from_slice(&orig_segs[squash_to + 1..]);
+        }
+        #[cfg(any(feature = "std", test))]
+        for s in &orig_segs[..squash_to + 1] {
+            mock_freeb(s.mp);
+        }
+
+        let mut off = 0;
+        info.offsets.inner.ether.pkt_pos = 0;
```

Comment:
This doesn't account for any outer headers

Reply:
We should only hit outer headers on inbound, in which case we won't hit the `shrink_start` path, so squashing does not matter. I've added a check to avoid squashing on the inbound case, so we don't need to worry about shifting around outer headers.

Reply:
Actually, scratch that we do hit `shrink_start` when new headers are smaller than old headers ... sigh.

Reply:
hmm, true but would we hit that case? `new_hdr_len > old_hdr_len` shouldn't be true in the inbound case right

Reply:
but i guess still no guarantee all the headers would be in the first segment